rholang/examples/locker.rho

```diff
@@ -0,0 +1,301 @@
+        lookup!(voterURI, *vlCh) |
+        for(voterLocker <- vlCh) {
+          voterLocker!(sig, 0,
+            "voteFor", "Rick", *eject, *ack)
+        }
+      }
+    } |
+
+    // Make and register chair locker; return URI.
+    contract chair(@"makeLocker", @nickname, @pubKey, @choices, return) = {
+      new seedEntry, lockerCh, lookupCh, ackCh in {
+        // ISSUE: how to manage registry URIs in eval mode?
+        lookup!(`rho:id:6s5xtg5d78a3qz9sp1nimqefhggznycwjwjyesij3fy9c6au7up9dp`,
+          *lookupCh) | for(Ballot <- lookupCh) {
+
+          Locker!(nickname, pubKey,
+
+            // Note we're _sending_ this ballot seed code, not running it.
+            for(return <- seedEntry) {
+              new bCh, self in {
+                Ballot!(choices, *bCh) | for(chair, winner <- bCh) {
+
+                  // These are 0-ary and total, but they fit in
+                  // the 1-ary partial gateway
+                  contract self(@"giveRightToVote", _, ej, return) = {
+                    chair!("giveRightToVote", *return)
+                  } |
+
+                  contract self(@"getWinner", _, ej, return) = {
+                    winner!(*return)
+                  }
+                } |
+                return!(bundle+{*self})
+              }
+            },
+            *seedEntry,
+            *lockerCh)
+        } |
+
+        for (@chairLocker <- lockerCh) {
+          insertArbitrary!(bundle+{chairLocker}, *return)
+        }
+      }
+    } |
+
+    contract voter(@"makeLocker", @nickname, @pubKey, return) = {
+      // trace!({"voter MakeLocker": nickname}) |
+      new seedEntry, lockerCh, inboxUriCh, ch in {
+        Locker!(nickname, pubKey,
+
+          // Note we're _sending_ this voter seed code, not running it.
+          for(return <- seedEntry) {
+            new self, rCh, inbox in {
+              insertArbitrary!(bundle+{*inbox}, *rCh) |
+              for(@uri <- rCh) {
+                inboxUriCh!(uri) |
+
+                contract self(@"voteFor", @choice, ej, ack) = {
+                  for(rightToVote <- inbox) {
+                    rightToVote!("vote", choice, *ej, *ack)
+                  }
+                }
+              } |
+              return!(bundle+{*self})
+            }
+          },
+          *seedEntry,
+          *lockerCh) |
+        for (voterLocker <- lockerCh) {
+          insertArbitrary!(bundle+{*voterLocker}, *ch) |
+          for(@lockerURI <- ch; @inboxURI <- inboxUriCh) {
+            trace!({"made voter locker": lockerURI, "inbox": inboxURI}) |
+            return!({"locker": lockerURI, "inbox": inboxURI})
+          }
+        }
+      }
+    }
+  } |
+  // end test / demo
+
+  new uriCh in {
+    insertArbitrary!(bundle+{*Locker}, *uriCh) | for(@uri <- uriCh) {
+      stdout!({"Locker registry uri": uri})
+    }
+  }
+  |
+
+  // WARNING: if sig is the wrong size, we get a java.lang.RuntimeException,
+  // and our nonceCh is stuck.
+  contract verify(
```

Comment:
I don't understand the need for the nonce here. 

The nonce would be a token used to avoid reusing the same signature twice. In our case the signed data includes the return channel, which is an unforgeable name. The `Locker` will send its response to this ret channel so nobody else will have access to that data.

So in this case we don't really need to worry about replay attacks.

Reply:
After the discussions in #cryptography discord channel I'm convinced that we do need the nonce. Still, I don't like the current implementation because the management of the nonces will be pretty complicated.

I have the following alternative proposal for the Locker contract (probably not syntactic correct):

```
contract Locker(
    @{label /\ String},
    @{pubKey /\ ByteArray},
    @seedCode, seedEntry,
    return) = {
    new self, seedCh in {
      return!(*self) |
      
      seedCode |
      seedEntry!(*seedCh) | for (seed <- seedCh) {
        contract self(@"getSeedProxy", @{sig /\ ByteArray}, return) = { 
          verify!(sig, pubKey, ("getSeedProxy", *return), *eject, *okCh) |
          for(_ <- okCh) {
            new proxy in {
              for (msg <- proxy) { seed ! (msg)} |
              return ! (proxy)
            }
          }
        }
      }
    }
  }
```

And now the Locker would be used like this:
```
Locker!("getSeedProxy", signature, return) | 
for (proxy <- return) { 
  // use the proxy
}```

Reply:
@JoshOrndorff in this morning's discussion you gave some feedback on this code... were you saying that this replay prevention technique is not OK? It seems OK to me.

Modulo the kludge where  I split msg into 3 parts, which should be left to a separate layer, I think the code I wrote recently pretty much matches this sketch.
https://github.com/rchain/rchain/blob/fe3a8cd65f6da64810a6f1be3e144d1c7bb7efd2/casper/src/main/rholang/MakeRev.rho#L208-L269

Thinking again about splitting msg or not... the `seed ! (msg)` pattern only works for 1-ary seed contracts. So it wouldn't work with stuff like `purse!("getBalance", *ch)`, right? let alone `purse!("deposit", src, amount, return)`.

Reply:
My realization during your office hours was that it meant keeping track of all unforgeable names ever used somehow. That is guaranteed at the blockchain level and is not the problem of the contract. This will work fine once the assumption that the same unforgeable name cannot be generated twice holds.

> only works for unary contracts
I know that's a big bummer. It's why I've added "make all contracts unary by explicit tupling" to the librho style guide.

> Wouldn't work with stuff like `purse!("getBalance", *ch)`, right?
That's right. It's easier to make design changes to the purse now than it will be in the future.

FWIW, I've convinced myself that the best pattern is contracts called on compound names with args passed in via explicit tupling.